File: src/backend/services/ChequeService.ts
import { db } from '../data/AppDbContext.ts';
import { Cheque } from '../models/entities.ts';

export class ChequeService {
  static getCheques(status?: string) {
    let query = `
      SELECT c.*, i.number as invoice_number, s.name as shop_name
      FROM cheques c
      JOIN invoices i ON c.invoice_id = i.id
      JOIN shops s ON i.shop_id = s.id
      WHERE 1=1
    `;
    const params: any[] = [];

    if (status) {
      query += ' AND c.status = ?';
      params.push(status);
    }

    query += ' ORDER BY c.due_date ASC';

    return db.prepare(query).all(...params);
  }

  static updateChequeStatus(id: number, data: { status: string, returnReason?: string, clearedDate?: string }) {
    const transaction = db.transaction(() => {
      const cheque = db.prepare('SELECT * FROM cheques WHERE id = ?').get(id) as Cheque;
      if (!cheque) throw new Error('Cheque not found');

      // Update cheque
      db.prepare(`
        UPDATE cheques
        SET status = ?, return_reason = ?, cleared_date = ?, updated_at = CURRENT_TIMESTAMP
        WHERE id = ?
      `).run(data.status, data.returnReason || null, data.clearedDate || null, id);

      // If status is paid, we might want to check if the invoice is now settled?
      // Actually, in the spec, "Paid" is already recorded in the invoice total.
      // The cheque record status just tracks the physical cheque clearance.
      // But if it's returned, we should subtract from invoice paid?
      // Spec says: "Invoice: set paid = sum(payments.amount)" at create time.
      // If cheque returns, invoice should technically go back to credit?
      // Let's implement that logic if it's 'returned'.

      if (data.status === 'returned') {
        const invoice = db.prepare('SELECT * FROM invoices WHERE id = ?').get(cheque.invoice_id) as any;
        const newPaid = invoice.paid - cheque.amount;
        db.prepare(`
          UPDATE invoices
          SET paid = ?, status = 'credit', updated_at = CURRENT_TIMESTAMP
          WHERE id = ?
        `).run(newPaid, cheque.invoice_id);
      }
    });

    transaction();
  }
}

File: backend/Services/ChequeService.cs
using Microsoft.EntityFrameworkCore;
using DistributionApi.Data;
using DistributionApi.DTOs;
using DistributionApi.Models;

namespace DistributionApi.Services;

public class ChequeService(AppDbContext db)
{
    public async Task<List<ChequeResponse>> GetChequesAsync(string? status)
    {
        var q = db.Cheques
            .Include(c => c.Invoice).ThenInclude(i => i.Shop)
            .AsQueryable();

        if (!string.IsNullOrEmpty(status))
            q = q.Where(c => c.Status == status);

        var list = await q.OrderByDescending(c => c.DueDate).ToListAsync();
        return list.Select(c => new ChequeResponse(
            c.Id, c.InvoiceId, c.ChequeNo, c.Bank, c.Amount,
            c.DateReceived, c.DueDate, c.ClearedDate, c.Status,
            c.ReturnReason, c.UpdatedAt,
            c.Invoice?.Number, c.Invoice?.Shop?.Name)).ToList();
    }

    public async Task<ChequeResponse?> UpdateChequeAsync(int id, UpdateChequeRequest req)
    {
        var cheque = await db.Cheques
            .Include(c => c.Invoice).ThenInclude(i => i.Shop)
            .FirstOrDefaultAsync(c => c.Id == id);

        if (cheque is null) return null;

        cheque.Status = req.Status;
        if (req.Status == "returned") cheque.ReturnReason = req.ReturnReason;
        if (req.Status == "paid") cheque.ClearedDate = req.ClearedDate ?? DateOnly.FromDateTime(DateTime.Today);
        cheque.UpdatedAt = DateTimeOffset.UtcNow;

        await db.SaveChangesAsync();

        return new ChequeResponse(
            cheque.Id, cheque.InvoiceId, cheque.ChequeNo, cheque.Bank, cheque.Amount,
            cheque.DateReceived, cheque.DueDate, cheque.ClearedDate, cheque.Status,
            cheque.ReturnReason, cheque.UpdatedAt,
            cheque.Invoice?.Number, cheque.Invoice?.Shop?.Name);
    }
}